### PR TITLE
[loss refactor] [1] snapshot tests + file structure

### PR DESCRIPTION
**Pull request by yueming-yuan**

### CODE REVIEW

File: miles/backends/training_utils/loss_hub/advantages.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+from miles.utils.distributed_utils import distributed_masked_whiten
+from miles.utils.ppo_utils import (
+    get_advantages_and_returns_batch,
+    get_grpo_returns,
+    get_reinforce_plus_plus_baseline_advantages,
+    get_reinforce_plus_plus_returns,
+)
+
+from ..cp_utils import get_logits_and_tokens_offset_with_cp
```
Comment (Shi-Dong) [inline miles/backends/training_utils/loss_hub/advantages.py:13]: Seems that many files in `training_utils` are using relative imports, but since this is a refactor, I'd suggest tightening up the convention of using absolute imports (as absolute import is strongly recommended by the Python style guide).

File: miles/backends/training_utils/loss_hub/advantages.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+def normalize_advantages(
+    args,
+    parallel_state,
+    advantages,
+    loss_masks,
+    total_lengths,
+    response_lengths,
+    max_seq_lens=None,
+):
```
Comment (Shi-Dong) [inline miles/backends/training_utils/loss_hub/advantages.py:103]: Nit: it'd be great to have input and return type annotations, as well as a docstring for this function.

File: miles/backends/training_utils/loss_hub/advantages.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    cp_size = parallel_state.cp_size
+    if cp_size == 1:
+        all_masks = torch.cat(loss_masks)
+    else:
+        mask_chunks = []
+        for i in range(len(advantages)):
+            total_len = total_lengths[i]
+            response_len = response_lengths[i]
+            prompt_len = total_len - response_len
+            max_seq_len = max_seq_lens[i] if max_seq_lens is not None else None
```
Comment (Shi-Dong) [inline miles/backends/training_utils/loss_hub/advantages.py:115]: I think it'd be safer to assert that `advantages`, `response_lengths`, `total_lengths`, and `max_seq_lens` all have the same length. Also here it'd be more Pythonic to use `zip` and `enumerate`.

File: miles/backends/training_utils/loss_hub/advantages.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+            _, _, _, token_offsets = get_logits_and_tokens_offset_with_cp(
+                total_len, response_len, parallel_state, args.qkv_format, max_seq_len
+            )
+
+            # Convert global offsets to response-space offsets
+            s0, e0 = token_offsets[0]
+            s1, e1 = token_offsets[1]
```
Comment (Shi-Dong) [inline miles/backends/training_utils/loss_hub/advantages.py:123]: Style-wise I'd prefer `(s0, e0), (s1, e1) = token_offsets` because this enforces that `token_offsets` is a 2-tuple of 2-tuples.

File: miles/backends/training_utils/loss_hub/corrections.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+def icepop_function(
+    args,
+    *,
+    pg_loss: torch.Tensor,
+    train_log_probs: list[torch.Tensor],
+    rollout_log_probs: list[torch.Tensor],
+    loss_masks: list[torch.Tensor],
+    **kwargs: Any,
+) -> tuple[torch.Tensor, list[torch.Tensor], dict[str, torch.Tensor]]:
```
Comment (Shi-Dong) [inline miles/backends/training_utils/loss_hub/corrections.py:38]: Same as above.

File: miles/backends/training_utils/loss_hub/losses.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+
+    return (
+        loss,
+        {
+            "loss": loss.clone().detach(),
+        },
+    )
+
+
+def get_loss_function(args) -> Callable:
```
Comment (Shi-Dong) [inline miles/backends/training_utils/loss_hub/losses.py:482]: I'd suggest tightening up the return type annotation: specifying the input and output types of `Callable`.  It'd be even better to define a `Protocol` for loss functions, just for type annotation purposes.

File: miles/backends/training_utils/loss.py [hunk 7]
```diff
@@ -904,11 +141,7 @@ def loss_function(
     else:
         loss, log = func(args, parallel_state, batch, logits, sum_of_sample_mean)
 
-    # With allgather-CP, some CP ranks may have no loss-contributing tokens (e.g., all
-    # padding). Without this, gradient doesn't flow through their attention path, so
-    # the CP gather's backward (reduce-scatter) is not called, deadlocking other CP
-    # ranks that call it. Adding this zero loss forces autograd to traverse the full
-    # graph on every rank without changing gradient values.
+    # Forces autograd to traverse the full graph on every rank to avoid hang.
```
Comment (Shi-Dong) [inline miles/backends/training_utils/loss.py:139]: I'm not sure why the comments are compressed into one line here. Doesn't seem to be related to the refactor?

File: miles/backends/training_utils/loss.py
Comment (Shi-Dong) [inline miles/backends/training_utils/loss.py]: Seems that the function also handles on-policy distillation?

**Review by Shi-Dong — COMMENTED** [review]
Refactor OK with minor comments.  Please fix all relative imports.

File: miles/backends/training_utils/loss.py (hunk 7 above)
Comment (yueming-yuan) [reply miles/backends/training_utils/loss.py:139]: The previous comment was also created by me, and I thought that one was too long when I looked at the code, so I shortened it randomly...

**Review by Zhichenzzz — APPROVED** [review]
LGTM! Thanks @yueming-yuan and @Shi-Dong
